diff -ruN include/floatp/f_d_floatp/d_float_dec2floatp.m include/floatp_mod/f_d_floatp/d_float_dec2floatp.m
--- include/floatp/f_d_floatp/d_float_dec2floatp.m	2021-09-14 16:14:58.000000000 +0100
+++ include/floatp_mod/f_d_floatp/d_float_dec2floatp.m	2022-02-10 06:24:42.795767336 +0000
@@ -62,14 +62,19 @@
  end % if
  
  [f,ex] = log2(y); % x = f 2^ex, 0.5 <= f <1, y < 1e308!!
- [fb,cnext] = f_d_frac2bin(f,2*nbits,sig); 
+ [fb,cnext] = f_d_frac2bin(f,53,sig);
  ind = find(fb); % find the position of the first 1
  
  if ~isempty(ind)
   I = 1;
   F = fb(ind(1)+1:end);
-  [F,cnext] = f_d_round_bin(F,nbits,sig); % round F to nbits 
   e = ex - ind(1);
+  if (e >= min_expo)
+    [F,cnext] = f_d_round_bin(F,nbits,sig); % round F to nbits
+  elseif e > (min_expo-nbits)
+    [F,cnext] = f_d_round_bin(F,nbits-(min_expo-e),sig); % round F to subnormal
+  end % if
+  e = e + cnext;
  else
   F = zeros(1,nbits,'uint8');
   e = ex;
@@ -83,7 +88,7 @@
   I = Inf(1,length(I)); 
   F = Inf(1,length(F));
  end % if
- if e < min_expo % underflow
+ if e < (min_expo-nbits) % underflow
   I = zeros(1,length(I));
   F = zeros(1,length(F));
  end % if
diff -ruN include/floatp/f_d_floatp/f_d_round_bin.m include/floatp_mod/f_d_floatp/f_d_round_bin.m
--- include/floatp/f_d_floatp/f_d_round_bin.m	2021-09-13 18:38:58.000000000 +0100
+++ include/floatp_mod/f_d_floatp/f_d_round_bin.m	2022-02-10 06:32:47.490152914 +0000
@@ -36,7 +36,7 @@
 
 len = length(bin);
 binr = bin(nbits+1:len);
-if len <= nbits 
+if len <= nbits | binr == 0
  % do nothing
  p = bin;
  cnext = 0;
